perma_web/perma/tasks.py

```diff
@@ -265,35 +386,316 @@ def makefile(self, *args, **kwargs):
 
     return requests_response
 
-def run_in_frames(browser, func, output_collector=None):
-    # setup
-    browser.implicitly_wait(0)
-    if output_collector is None:
-        output_collector = []
 
-    run_in_frames_recursive(browser, func, output_collector)
+### CAPTURE COMPONENTS ###
+
+# x-robots headers
+
+def xrobots_blacklists_perma(robots_directives):
+    darchive = False
+    if robots_directives:
+        for directive in robots_directives.split(";"):
+            parsed = directive.lower().split(":")
+            # respect tags that target all crawlers (no user-agent specified)
+            if len(parsed) == 1:
+                if "noarchive" in parsed:
+                    darchive = True
+            # look for perma user-agent
+            elif len(parsed) == 2:
+                if parsed[0] == "perma" and "noarchive" in parsed[1]:
+                    darchive = True
+            # if the directive is poorly formed, do our best
+            else:
+                if "perma" in directive and "noarchive" in directive:
+                    darchive = True
+    return darchive
 
-    # reset
-    browser.implicitly_wait(ELEMENT_DISCOVERY_TIMEOUT)
-    browser.switch_to.default_content()
+# meta tags
 
-    return output_collector
+def meta_thread(browser=None, link=None):
+    meta_tags = meta_tags_get(browser, link)
+    if meta_tags:
+        meta_tags_process(meta_tags, link)
 
-def run_in_frames_recursive(browser, func, output, frame_path=None):
-    if frame_path is None:
-        frame_path = []
-    output += func(browser)
-    for child_frame in browser.find_elements_by_tag_name('frame') + browser.find_elements_by_tag_name('iframe'):
-        browser.switch_to.default_content()
-        for frame in frame_path:
-            browser.switch_to.frame(frame)
+def meta_tags_get(browser, link):
+    """
+        Retrieves meta tags. Returns a list of dicts, keys=name, content.
+        Uses js instead of selenium for speed.
+    """
+    def get_meta():
+        with browser_running(link, browser, lambda: meta_tag_analysis_failed(link)):
+            return browser.execute_script("""
+                var meta_tags = document.getElementsByTagName('meta');
+                var tags = [];
+                for (var i = 0; i < meta_tags.length; i++){
+                    tags.push({"name":meta_tags[i].name, "content":meta_tags[i].content});
+                }
+                return tags
+            """)
+    return repeat_until_truthy(get_meta)
+
+def meta_tags_process(meta_tags, link):
```

Comment:
Sure! So long as we are comfortable collapsing duplicate names. Since duplicates aren't allowed for the meta tags we are currently looking at, seems totally reasonable. Why go with the first instead of the last though? Any reason to think an improperly coded page would intend for us to respect the first, rather than the last, robots declaration, etc.?